Centralize setting messageId in RequestCallback

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* **#57848 Centralize setting messageId in RequestCallback**
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

lw committed:
Centralize setting messageId in RequestCallback

This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]

lw committed:
Update on "Centralize setting messageId in RequestCallback"


This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]

lw committed:
Update on "Centralize setting messageId in RequestCallback"


This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -481,7 +459,8 @@ c10::intrusive_ptr<JitFuture> RequestCallbackImpl::processRpcWithErrors(
     return future;
   } catch (std::exception& e) {
-    return asFuture(handleError(e, messageType, messageId));
+    // Pass a dummy message ID since it will be overwritten anyways.
```
should we remove message id from  `handleError` argument list?

mrshenli replied:
Please feel free to do this in followup PRs.

lw replied:
My long-term idea is to remove the ID entirely from Message. I believe the ID is only used by the RPC agent to match requests with responses, so in a sense it's an "external" information (it shouldn't be _inside_ the Message, it should be a "tag" that's attached to it).

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -109,7 +108,16 @@ c10::intrusive_ptr<JitFuture> RequestCallbackNoPython::processMessage(
+          message->setId(id);
+          return message;
+        },
+        c10::getCustomClassType<c10::intrusive_ptr<Message>>());
```
IIUC, this does some type conversion + map lookup + branching. I wonder if it make sense to keel this `c10::ClassTypePtr` in Message implementation and reuse it in other places?

lw replied:
I wouldn't worry about performance here. It's true that we need to do a map lookup and some bookkeeping, but only the first time, as the result is then cached for subsequent uses. See the `static` local variable here https://github.com/pytorch/pytorch/blob/df44f015fe1403f0fea54b46736c07550bb85870/aten/src/ATen/core/ivalue.h#L1161-L1169

Moreover, since that function is templated, the type will be resolved at compile time. Hence a reasonable compiler will inline that code and this callsite will basically just become one memory access (plus a check if the cache has been initialized).

mrshenli reviewed (approved):
LGTM

lw committed:
Update on "Centralize setting messageId in RequestCallback"


This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]

lw committed:
Update on "Centralize setting messageId in RequestCallback"


This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]

lw committed:
Update on "Centralize setting messageId in RequestCallback"


This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]

lw committed:
Update on "Centralize setting messageId in RequestCallback"


This PR looks large, but all it does is add a dozen lines and remove a lot of other ones.

One first advantage of using Futures is that we can easily chain some "post-processing" to them. Until now we needed to pass the message ID around everywhere because it was set separately by each method. Instead, we could simply add a follow-up step to the final future which sets this ID, and remove all the former logic.

Differential Revision: [D28224477](https://our.internmc.facebook.com/intern/diff/D28224477/)

[ghstack-poisoned]
